### PR TITLE
memutils: memmove() / memcpy() implementation

### PR DESCRIPTION
**Pull request by baziotis**
Notes:
1) I could not find a better way to handle both static and dynamic arrays (that includes `isArray`).
2) More tests to be added.
3) To check benchmarks, I recommend the main repo: https://github.com/baziotis/Dmemutils

### CODE REVIEW

File: src/core/experimental/memutils.d [hunk 1]
```diff
@@ -0,0 +1,685 @@
...
+    }
+    else if ((cast(ulong)s - cast(ulong)d) < n)
+    {  // There is overlap with src being ahead. Use backwards move.
+        Dmemmove_forw(d, s, n);
+    }
+    else
+    {  // There is no overlap, use memcpy.
+        pragma(inline, true);
+        Dmemcpy(d, s, n);
+    }";
```
Comment (wilzbach) [inline src/core/experimental/memutils.d]: Private. Also, what's the motivation for the split-up? Why can't the static array implementation call the other one?

File: src/core/experimental/memutils.d [hunk 2]
```diff
@@ -0,0 +1,685 @@
...
+    {  // There is overlap with src being ahead. Use backwards move.
+        Dmemmove_forw(d, s, T.sizeof);
+    }
+    else
+    {  // There is no overlap, use memcpy.
+        Dmemcpy(dst, src);
+    }
+}
+
+unittest
```
Comment (wilzbach) [inline src/core/experimental/memutils.d]: This needs `///` in order to be part of the documentation.

File: src/core/experimental/memutils.d [hunk 3]
```diff
@@ -0,0 +1,685 @@
...
+    const float[3] a = [1.2, 3.4, 5.8];
+    float[3] b;
+    memmove(b, a);
+    assert(b[0] == 1.2f);
+    assert(b[1] == 3.4f);
+    assert(b[2] == 5.8f);
+}
+
+/* Can we use SIMD?
+ */
```
Comment (wilzbach) [inline src/core/experimental/memutils.d]: This comment isn't really helpful.

File: src/core/experimental/memutils.d [hunk 4]
```diff
@@ -0,0 +1,685 @@
...
+        Dmemmove_forw(d, s, n);
+    }
+    else
+    {  // There is no overlap, use memcpy.
+        pragma(inline, true);
+        Dmemcpy(d, s, n);
+    }";
+
+
+
```
Comment (wilzbach) [inline src/core/experimental/memutils.d]: Only one line of whitespace between functions.

File: src/core/experimental/memutils.d [hunk 5]
```diff
@@ -0,0 +1,685 @@
...
+    assert(b[0] == 1.2f);
+    assert(b[1] == 3.4f);
+    assert(b[2] == 5.8f);
+}
+
+/* Can we use SIMD?
+ */
+version (D_SIMD)
+{
+    import core.simd: float4;
```
Comment (wilzbach) [inline src/core/experimental/memutils.d]: DStyle: ` : `

File: src/core/experimental/memutils.d [hunk 6]
```diff
@@ -0,0 +1,685 @@
+/**
+ * Pure D replacement of the C Standard Library basic memory building blocks of string.h
+ * Source: $(DRUNTIMESRC core/experimental/memutils.d)
+ */
+module core.experimental.memutils;
+
+void memmove(T)(T *dst, const T *src)
```
Comment (wilzbach) [inline src/core/experimental/memutils.d]: Needs full Ddoc comment.
Also, what's the reason for not reason `ref`?

Comment (wilzbach) [reply src/core/experimental/memutils.d]: Also, thinking more about this interface `src, dst` would be more D-like, but I guess we can't really change that.

File: src/core/experimental/memutils.d [hunk 7]
```diff
@@ -0,0 +1,685 @@
...
+    real b;
+    memmove(&b, &a);
+    assert(b == 1.2);
+    // Overwrite the type system and create overlap with dst forward.
+    ubyte[8] buf;
+    int *p = cast(int*) buf.ptr;
+    int *q = cast(int*) (buf.ptr + 2);
+    *p = 203847;
+    memmove(q, p);
+    assert(*q == 203847);
```
Comment (wilzbach) [inline src/core/experimental/memutils.d]: Split up into multiple tests, no?

File: src/core/experimental/memutils.d [hunk 8]
```diff
@@ -0,0 +1,685 @@
...
+    assert(b[1] == 3.4f);
+    assert(b[2] == 5.8f);
+}
+
+/* Can we use SIMD?
+ */
+version (D_SIMD)
+{
+    import core.simd: float4;
+    enum useSIMD = true;
```
Comment (wilzbach) [inline src/core/experimental/memutils.d]: Needs to be private.

File: src/core/experimental/memutils.d [hunk 9]
```diff
@@ -0,0 +1,685 @@
...
+    else version (DigitalMars)
+    {
+        import core.simd: void16, loadUnaligned, storeUnaligned;
+    }
+    else version (GNU)
+    {
+        import gcc.builtins : __builtin_ia32_storeups, __builtin_ia32_loadups;
+    }
+
+    void store16f_sse(void *dest, float4 reg)
```
Comment (wilzbach) [inline src/core/experimental/memutils.d]: Private

File: src/core/experimental/memutils.d [hunk 10]
```diff
@@ -0,0 +1,685 @@
...
+        enum useSIMD = false;
+    }
+}
+else
+{
+    enum useSIMD = false;
+}
+
+/* Little SIMD library
+ */
```
Comment (wilzbach) [inline src/core/experimental/memutils.d]: Maybe move into a separate module?

File: src/core/experimental/memutils.d [hunk 11]
```diff
@@ -0,0 +1,685 @@
...
+        // Move pointers to their start.
+        s -= n;
+        d -= n;
+        goto START;
+    }
+}
+
+/* Handle dynamic sizes < 64 with forwards move. Overlap is possible.
+ */
+void Dmemmove_forw_lt64(void *d, const(void) *s, size_t n)
```
Comment (wilzbach) [inline src/core/experimental/memutils.d]: Private

File: src/core/experimental/memutils.d [hunk 12]
```diff
@@ -0,0 +1,685 @@
...
+        else version (DigitalMars)
+        {
+            storeUnaligned(cast(void16*)dest, reg);
+        }
+        else version (GNU)
+        {
+            __builtin_ia32_storeups(cast(float*) dest, reg);
+        }
+    }
+    float4 load16f_sse(const(void) *src)
```
Comment (wilzbach) [inline src/core/experimental/memutils.d]: No underscore - see the DStyle.

File: src/core/experimental/memutils.d (hunk 1 above)
Comment (baziotis) [reply src/core/experimental/memutils.d]: The most important reason is that I couldn't make `(ref T[] dst, ref const T[] src)` work when the arguments are `!()(float[3], const(float[3]))`. I don't know why that is.

File: src/core/experimental/memutils.d [hunk 13]
```diff
@@ -0,0 +1,685 @@
+/**
+ * Pure D replacement of the C Standard Library basic memory building blocks of string.h
+ * Source: $(DRUNTIMESRC core/experimental/memutils.d)
+ */
+module core.experimental.memutils;
+
+void memmove(T)(T *dst, const T *src)
+{
+    void *d = cast(void*) dst;
```
Comment (thewilsonator) [inline src/core/experimental/memutils.d]: please write `void* d` not `void *d`, this is D not C. ditto throughout.

File: src/core/experimental/memutils.d [hunk 14]
```diff
@@ -0,0 +1,685 @@
...
+}
+
+/* Static Arrays
+ */
+void memmove(T, size_t len)(ref T[len] dst, ref const T[len] src)
+{
+    mixin(arrayCode);
+}
+
+enum arrayCode = "
```
Comment (thewilsonator) [inline src/core/experimental/memutils.d]: if you're going to use enum+mixin, please use `q{}` strings.

File: src/core/experimental/memutils.d (hunk 10 above)
Comment (thewilsonator) [reply src/core/experimental/memutils.d]: one possibility is to make `core.experimental.mem` a package with modules `cpy`, `move`, and `simd`.

Comment (wilzbach) [reply src/core/experimental/memutils.d]: Better name it `memory` as there's `core.memory`.

Comment (baziotis) [reply src/core/experimental/memutils.d]: Ok, I'm making a new PR. I have seen all the comments, I left them to be resolved in the new PR.

**Pull request closed without merging**